### PR TITLE
Blue typescript

### PR DESCRIPTION
**Pull request by skqksh**
## Proposed changes

- Support Typescript (#157)

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/caver-js/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/caver-js)
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

-#157

## Further comments

I added declare files to support typescript. I focused to change files where are in official docs
Lots of testcases include unnecessary options(parameters, etc) for the test, so I hope to change *.js files to *.ts
I believe that It will help to find bugs or prevent other type errors

### CODE REVIEW

**Comment by skqksh** [comment]
@jimni1222 

Awesome ~! Thanks for your help

boo

**Comment by jimni1222** [comment]
@skqksh 

Yeahhh~ now this PR has passed the circle ci test. 👍 

Can i ask you something?
What is the criteria for selecting files converted to ts?

And as you can see in the [common architecture](http://kips.klaytn.com/KIPs/kip-34), Caver's interface naming rule is ʻIXXX`. Please change the interface name from `XXX_I` to ʻIXXX`.

**Comment by skqksh** [comment]
@jimni1222 

I changed interface's name

- What is the criteria for selecting files converted to ts?
I converted files written in Klaytn-Docs, first. Some codes seem not using anymore
I didn't convert some files which are still alive in caver-js. I just remained that for someone who want to try(Maybe me of the future?)
I hope you guys to change other testcases to .ts, It must be good to make nicer API

File: packages/caver-account/src/accountKey/accountKeyPublic.d.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
+/**
+ * Representing an AccountKeyPublic.
+ * @class
+ */
+export interface IAccountKeyPublic {
```
Comment (jimni1222) [inline packages/caver-account/src/accountKey/accountKeyPublic.d.ts:5]: Why should interface and class be defined in d.ts file at the same time? What purpose is each used for?

**Comment by jimni1222** [comment]
@skqksh Please add the license text at the top of each newly created file. One blank line is required after the license text.

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+     */
+    static providers: any
+    /**
+     * @deprecated it is not official property
+     */
+    static abi: any
+
+    constructor(provider: any, net?: any)
+
+    account: IAccount
```
Comment (jimni1222) [inline index.d.ts:28]: Why are some types defined as interfaces? By what criteria is it divided?

File: tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    // "alwaysStrict": true,                  /* Parse in strict mode and emit "use strict" for each source file. */
+
+    /* Additional Checks */
+    // "noUnusedLocals": true,                /* Report errors on unused locals. */
+    // "noUnusedParameters": true,            /* Report errors on unused parameters. */
+    // "noImplicitReturns": true,             /* Report error when not all code paths in function return a value. */
+    // "noFallthroughCasesInSwitch": true,    /* Report errors for fallthrough cases in switch statement. */
+
+    /* Module Resolution Options */
+     "moduleResolution": "node",            /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
```
Comment (jimni1222) [inline tsconfig.json:44]:
```suggestion
    "moduleResolution": "node",            /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
```

File: test/packages/caver.account.ts [hunk 4]
```diff
@@ -1,50 +1,46 @@
...
-    (at your option) any later version.
-
-    The caver-js library is distributed in the hope that it will be useful,
-    but WITHOUT ANY WARRANTY; without even the implied warranty of
-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
-    GNU Lesser General Public License for more details.
-
-    You should have received a copy of the GNU Lesser General Public License
-    along with the caver-js. If not, see <http://www.gnu.org/licenses/>.
-*/
```
Comment (jimni1222) [inline test/packages/caver.account.ts:17]: Please don't remove license comments

File: packages/caver-account/src/accountKey/accountKeyPublic.d.ts (hunk 1 above)
Comment (skqksh) [reply packages/caver-account/src/accountKey/accountKeyPublic.d.ts:5]: @jimni1222 
yes. I made it.

To make instance of class in as a class property, I had to define interface
You can refer to [Typescript Docs](https://www.typescriptlang.org/docs/handbook/interfaces.html#class-types)
For example, if you set type as class(not interface), it returns error
```
export class Caver {
    ...
    account: Account // not account: IAccount
    ...
}
```
`
This expression is not constructable.
  Type 'Account' has no construct signatures.
`

**Comment by skqksh** [comment]
> @skqksh Please add the license text at the top of each newly created file. One blank line is required after the license text.

okay~

File: index.d.ts (hunk 2 above)
Comment (skqksh) [reply index.d.ts:28]: @jimni1222 
the reason is same with above

File: test/packages/caver.account.ts (hunk 4 above)
Comment (skqksh) [reply test/packages/caver.account.ts:17]: @jimni1222 

okay~

**Comment by devhyunjae** [comment]
@here Any progress in here? Need ts support!

**Comment by jimni1222** [comment]
@devhyunjae Sorry, but this PR is still under review.
@skqksh Please resolve the conflict.( and please rebase on latest dev branch :) ) And when will the request made in the comment above be reflected?

**Comment by skqksh** [comment]
@jimni1222 
I found that some of my codes(usage of  typescript) are wrong.
I'll rework it someday...
But now, Please close this PR. T^T

**Comment by jimni1222** [comment]
@skqksh Oh... I see! First, I will close this PR then. :) Thank you for always.

**Pull request closed without merging**